Fluid/Values/FluidValue.cs

```diff
@@ -167,7 +175,42 @@ public static FluidValue Create(object value, TemplateOptions options)
 
                         case FluidValue[] array:
                             return new ArrayValue(array);
+                    }
+
+                    // Check if it's a more specific IDictionary<string, V>, e.g. JObject
+
+                    if (!_genericDictionaryTypeCache.TryGetValue(typeOfValue, out var genericType))
+                    {
+                        lock (_genericDictionaryTypeCache)
+                        {
+                            if (!_genericDictionaryTypeCache.TryGetValue(typeOfValue, out genericType))
+                            {
+                                foreach (var i in typeOfValue.GetInterfaces())
+                                {
+                                    if (i.IsGenericType && i.GetGenericTypeDefinition() == typeof(IDictionary<,>) && i.GetGenericArguments()[0] == typeof(string))
+                                    {
+                                        genericType = typeof(ObjectDictionaryFluidIndexable<>).MakeGenericType(i.GetGenericArguments()[1]);
+
+                                        // Swap the previous cache with a new copy to prevent locking on reads
```

Comment:
is this swap really needed as lock is only taken on cache miss? we are also locking against the field here and the field is being changed

Reply:
I can agree with the lock instance, but not that this is an issue, just that using a dedicated lock object prevent unnecessary second check that could go null since now everything is waiting on the same object.

However I think the swap is necessary since the dictionary is static and could be accessed in read or write concurrently. So here a dictionary is always accessed either on read or on writes. If it's not swap then another thread you read it (first statement with the `TryGetValue`) while another one is adding an entry.